### PR TITLE
Mutation testing

### PR DESCRIPTION
**Pull request by psfinaki**
This:
- renames "mutation tests" to "mutation testing"
- removes the hack about JSON file rewriting, leverages "native" --target-framework cmd parameter in Stryker
- adds a preliminary version of JSON merging utility

### CODE REVIEW

File: eng/MSBuild/MutationTesting.props [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
   <!-- Stryker.NET doesn't support passing arguments to dotnet or MSBuild processes, so we have to set it explicitly -->
   <PropertyGroup>
-    <!-- Mutation tests only run for the latest TFM -->
+    <!-- Mutation testing onlys run for the latest TFM -->
```
Comment (RussKie) [inline eng/MSBuild/MutationTesting.props:5]:
```suggestion
    <!-- Mutation testing only run for the latest TFM -->
```

File: eng/build.ps1 [hunk 2]
```diff
@@ -61,17 +61,17 @@ Param(
...
+  [switch] $mutationTesting,
 
   [Parameter(ValueFromRemainingArguments=$true)][String[]]$properties
 )
 
 function Print-Usage() {
   Write-Host "Custom settings:"
   Write-Host "  -testCoverage           Run unit tests and capture code coverage information."
-  Write-Host "  -mutationTest           Run mutation tests."
+  Write-Host "  -mutationTesting           Run mutation testing."
```
Comment (RussKie) [inline eng/build.ps1:74]: Please make sure to valign the description with the rest

File: eng/build.sh [hunk 3]
```diff
@@ -11,7 +11,7 @@ usage()
 {
   echo "Custom settings:"
   echo "  --testCoverage             Run unit tests and capture code coverage information"
-  echo "  --mutationTest             Run mutation tests"
+  echo "  --mutationTesting             Run mutation testing"
```
Comment (RussKie) [inline eng/build.sh:14]: Ditto

**Review by RussKie — COMMENTED** [review]
Few minor comments

File: eng/pipelines/templates/MutationTesting.yml [hunk 4]
```diff
@@ -0,0 +1,49 @@
+parameters:
```
Comment (RussKie) [inline eng/pipelines/templates/MutationTesting.yml:1]: Don't forget to delete MutationTests.yml

**Comment by RussKie** [comment]
@psfinaki it looks like after I changed the target branch from "mutationtests" to "experimental/mutationtests" git included unrelated changes. Please do a clean rebase of your work. Thank you.

File: eng/StrykerNET/stryker-config.json
Comment (psfinaki) [inline eng/StrykerNET/stryker-config.json:1]: revise this before merging

File: eng/build.ps1 [hunk 5]
```diff
@@ -143,8 +143,15 @@ try {
...
-    Remove-Item -Path .mutationtests
+  if ($mutationTesting) {
+    Remove-Item -Path .mutationtesting
+
+    $testResultPath = "./artifacts/TestResults/$configuration";
+
+    # Merge JSON reports
+
+    # Open HTML report
+    Start-Process $testResultPath/..../mutation-report.html
```
Comment (RussKie) [inline eng/build.ps1:154]: Noting the TODOs here

File: eng/build.sh [hunk 6]
```diff
@@ -95,17 +95,17 @@ if [[ "$mutationTest" == true ]]; then
...
 fi
 
 "$DIR/common/build.sh" $properties
 
 # Remove the marker when we're done
-if [[ "$mutationTest" == true ]]; then
-  [ -e "$REPO_ROOT/.mutationtests" ] && rm -- "$REPO_ROOT/.mutationtests"
+if [[ "$mutationTesting" == true ]]; then
+  [ -e "$REPO_ROOT/.mutationtesting" ] && rm -- "$REPO_ROOT/.mutationtesting"
 fi
```
Comment (RussKie) [inline eng/build.sh:109]: Please add a comparable report merging functionality here as well. Similar to
https://github.com/Azure/dotnet-extensions-experimental/blob/de1c8a6606ebcf6266a6d7020a3a3820a0e1f635/eng/build.sh#L80-L84

Please note, that we can't open the report automatically.

Comment (psfinaki) [reply eng/build.sh:109]: Agreed offline this would be a followup

File: eng/build.ps1 (hunk 5 above)
Comment (psfinaki) [reply eng/build.ps1]: Yep, merging works as of now

**Comment by psfinaki** [comment]
@RussKie Okay this is green, the merged report seems to be generated correctly, here is [the example](https://dev.azure.com/dnceng-public/public/_build/results?buildId=355653&view=stryker-mutator.mutation-report-publisher.mutation-report-tab) from the latest build.

I believe this can go in and then we can build other functionality on top it.

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
@psfinaki I rolled back `/m:1` option that serialised all mutation tests, and now stryker is having a hard time...

**Comment by psfinaki** [comment]
@RussKie I see 2 options here then:
1. Switch off mutation testing for projects where Stryker has hard times (by removing the `ProjectUnderTest` thing) - those seem to be HealthChecks.ServiceFabric and Telemetry - and then we can investigate them separately.
2. Set up `/m:1` again and, well, investigate the problematic projects anyway so that we can remove `/m:1` later.

File: eng/StrykerNET/StrykerNET.targets [hunk 7]
```diff
@@ -34,14 +34,6 @@
   </Target>
 
   <Target Name="RunMutationTesting">
-
-    <PropertyGroup>
-      <_OnlyTfmPath>$([MSBuild]::NormalizePath('$(RepoRoot)', '.targetframeworks'))</_OnlyTfmPath>
-    </PropertyGroup>
-    <WriteLinesToFile File="$(_OnlyTfmPath)"
-                      Overwrite="true"
-                      Lines="$(LatestTargetFramework)" />
```
Comment (RussKie) [inline eng/StrykerNET/StrykerNET.targets]: Looks like this introduced a race condition. These targets are execute per test project, so while this file was (re-)written another project may have been started...

File: eng/build.ps1 [hunk 8]
```diff
@@ -109,6 +109,7 @@ if ($mutationTesting) {
 
   # Create a marker file
   '' | Out-File .mutationtesting
+  'net8.0' | Out-File .targetframeworks
```
Comment (RussKie) [inline eng/build.ps1:112]: This is kind of unfortunate that we have to hardcode it. But I guess it's a small price to pay.

Comment (psfinaki) [reply eng/build.ps1:112]: Yeah. Probably we can eventually probably the TFM info from the pipeline once anything else needs it.

**Comment by psfinaki** [comment]
@RussKie thanks for looking into this. Now it looks things work at least on Windows. 
I update the Stryker config to match the repo reality, otherwise LGTM.

**Pull request merged**